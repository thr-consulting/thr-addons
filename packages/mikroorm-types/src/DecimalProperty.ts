import {Type, ValidationError} from '@mikro-orm/core';

export class DecimalProperty extends Type {
	convertToDatabaseValue(value: any): any {
		try {
			if (!value) return value;
			if (!parseFloat(value)) throw new Error();
			return JSON.stringify(value);
		} catch (err) {
			throw ValidationError.invalidType(DecimalProperty, value, 'JS');
		}
	}

	convertToJSValue(value: string): any {
		try {
			if (!value) return value;
			const number = parseFloat(value);
			if (!number) throw new Error();
			return number;
		} catch (err) {
			throw ValidationError.invalidType(DecimalProperty, value, 'database');
		}
	}

	getColumnType() {
		return 'decimal';
	}
}
