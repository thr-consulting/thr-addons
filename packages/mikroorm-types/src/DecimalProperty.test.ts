import {DecimalProperty} from './DecimalProperty';

describe('DecimalProperty', () => {
	it('should convert Decimal values', () => {
		const decimal = 1.1;
		const prop = new DecimalProperty();

		expect(prop.convertToDatabaseValue(decimal)).toBe('1.1');
		expect(prop.convertToJSValue('2.2')).toBe(2.2);
		expect(prop.getColumnType()).toBe('decimal');
	});

	it('should throw errors on invalid values', () => {
		const decimal = 'not money';
		const prop = new DecimalProperty();

		function convertDbValue() {
			prop.convertToDatabaseValue(decimal);
		}
		function convertJsValue() {
			prop.convertToJSValue(decimal);
		}

		expect(convertDbValue).toThrowErrorMatchingSnapshot();
		expect(convertJsValue).toThrowErrorMatchingSnapshot();
	});
});
