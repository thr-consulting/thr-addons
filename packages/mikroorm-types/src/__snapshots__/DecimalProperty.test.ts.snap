// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DecimalProperty should throw errors on invalid values 1`] = `"Could not convert JS value 'not money' of type 'string' to type DecimalProperty"`;

exports[`DecimalProperty should throw errors on invalid values 2`] = `"Could not convert database value 'not money' of type 'string' to type DecimalProperty"`;
